reconcile/utils/saasherder.py

```diff
@@ -141,10 +145,29 @@ def _validate_saas_files(self):
                     # promotion publish channels
                     promotion = target.get('promotion')
                     if promotion:
+                        rt_ref = (saas_file_path,
+                                  resource_template_name,
+                                  resource_template_url)
+
                         publish = promotion.get('publish')
                         if publish:
-                            saas_file_promotion_publish_channels.extend(
-                                publish)
+                            for channel in publish:
+                                if channel in publications:
+                                    self.valid = False
+                                    logging.error(
+                                        "saas file promotion publish channel"
+                                        "is not unique: {}"
+                                        .format(channel)
+                                    )
+                                    continue
+                                publications[channel] = rt_ref
+
+                        subscribe = promotion.get('subscribe')
+                        if subscribe:
+                            for channel in subscribe:
+                                subscriptions.setdefault(channel, [])
+                                subscriptions[channel].append(rt_ref)
+
```

Comment:
The function is quite humongous already. Does it make sense to move this portion into a dedicated function, like

```
# input target
# output subscriptions[channel, list[]]
# output publications[channel, RTRef]
def _get_channels_from_pomotion(self, target: Mapping[str, Any]) -> dict[str, list[str]], dict[str, RTRef]
```

That way, we could also remove the comment `# promotion publish channels`

Probably, instead of having `rt_ref` passed around as tuple, we could use a dataclass

```
@dataclass
class RTRef:
    saas_file_path: str
    resource_template_name: str
    resource_template_url: str
```

Reply:
TBH I think this code needs a refactor. 
Separating this part into a function requires more that what you pointed out, for example, RT_Ref attributes are not in the object, are local to the function, etc. This PR intention is to add a validation to ensure saas specs are correct, not to improve the python code.

About the dataclass, I prefer to maintain the same style on the whole file and not introduce a new concept just for this part, I think it's better to have a model on this that covers the whole saas spec, not just a local ref.

Reply:
> This PR intention is to add a validation to ensure saas specs are correct, not to improve the python code.

Only concern is that in my experience dedicated refactor PRs never come. Usually what this means is that stuff just grows out of hand over time to a point were any refactor will be huge amount of work (looking at `terrascript_client.py` for instance :eyes: ). Personally, I feel like every **feature** PR's goal should also be to (slightly) improve the code it affects (if possible). That's why Im always in favor of making small enhancements as you touch a portion of code - and annoying people with it :sweat_smile: . E.g., in this case it would allow us to add some type safety via the function signatures on top of de-cluttering the method. 

That being said, thank you for bearing with me :) I might as well be too idealistic about this. As you said, the effort is bigger than what I initially anticipated and might not really be pragmatic right now.

Reply:
I completely agree with you. 

I'm going to separate the logic you mention. I won't add the dataclass bc I feel it needs a broader change scope, but I think it's worth to slim down this function a bit. 

You made me feel bad... 😝